core/include/vecmem/edm/data.hpp

```diff
@@ -0,0 +1,115 @@
+/// directly create a @c vecmem::edm::view object. Not if there are any jagged
+/// vectors in the container. As those require the data object to hold on to
+/// some memory. Exactly like @c vecmem::data::jagged_vector_data does.
+///
+/// This is the type created by
+/// @c vecmem::get_data(vecmem::edm::host<schema<VARTYPES...>>&) calls.
+/// Which then needs to be held on to by the host code, while a
+/// @c vecmem::edm::view<schema<VARTYPES...>> of the data object is passed to
+/// any device function that wants to use data from the host container.
+///
+/// @tparam ...VARTYPES The variable types to store in the data object
+///
+template <typename... VARTYPES>
+class data<schema<VARTYPES...>> : public view<schema<VARTYPES...>> {
+
+public:
+    /// The schema describing the buffer's payload
+    using schema_type = schema<VARTYPES...>;
+    /// Base view type
+    using view_type = view<schema_type>;
+    /// Size type used for the container
+    using size_type = typename view_type::size_type;
+    /// The tuple type holding all of the data objects for the individual
+    /// variables
+    using tuple_type =
+        std::tuple<typename details::data_type<VARTYPES>::type...>;
+
+    /// @name Constructors and assignment operators
+    /// @{
+
+    /// Default constructor
+    data() = default;
+
+    /// Constructor for the data object
+    ///
+    /// @param size The size of the (outer) arrays in the data object
+    /// @param resource The memory resource to use for the allocation(s)
+    ///
+    VECMEM_HOST
+    data(size_type size, memory_resource& resource);
+
+    /// @}
+
+    /// @name Function(s) meant for internal use by other VecMem types
+    /// @{
+
+    /// Direct (non-const) access to the underlying tuple of data objects
+    VECMEM_HOST
+    tuple_type& variables();
+    /// Direct (const) access to the underlying tuple of data objects
+    VECMEM_HOST
+    const tuple_type& variables() const;
+
+    /// @}
+
+private:
+    /// Variable holding data for the jagged vector variables
+    tuple_type m_data;
+
+};  // class data
+
+}  // namespace edm
+
+/// Helper function for getting a (possibly non-const) view of a data object
+///
+/// @tparam ...VARTYPES The variable types describing the container
+/// @param data The data object to get a view for
+/// @return A (possibly non-const) view into for the data object
+///
+template <typename... VARTYPES>
+VECMEM_HOST edm::view<edm::schema<VARTYPES...>>& get_data(
+    edm::data<edm::schema<VARTYPES...>>& data);
+
+/// Helper function for getting a (const) view of a data object
+///
+/// @tparam ...VARTYPES The variable types describing the container
+/// @param data The data object to get a view for
+/// @return A (const) view into for the data object
+///
+template <typename... VARTYPES>
+VECMEM_HOST const edm::view<edm::schema<VARTYPES...>>& get_data(
+    const edm::data<edm::schema<VARTYPES...>>& data);
```

Comment:
Let me pose one philosophical question to you... What should be the return type of this function?

For `vecmem::edm::host` we return a different C\+\+ type from `vecmem::get_data(const vecmem::edm::host&)` from the type returned by `vecmem::get_data(vecmem::edm::host&)`. But here I didn't do it.

The return type of this could also be:

```c++
edm::view<edm::details::add_const_t<edm::schema<VARTYPES...>>>
```

In fact, that's exactly what I did for the buffer:

https://github.com/krasznaa/vecmem/blob/EDMDevelopments-main-20231031/core/include/vecmem/edm/buffer.hpp#L153-L172

But I can't decide right now which one is better. Whether the forced type change might make something harder to do later on. 🤔 (Remember, a non-const view can always be converted into a constant one.)

Reply:
Decided to go with the constant type in the end.